beluga/include/beluga/localization.hpp

```diff
@@ -70,30 +83,60 @@ class NullMixin : public Mixin {
  * \tparam CustomExtensionMixinInterface An optional mixin interface for the user to customize the filter.
  */
 template <class Map, class CustomExtensionMixinInterface = NullMixinInterface>
-using LaserLocalizationInterface2d = beluga::mixin::compose_interfaces<
+using LaserLocalizationInterface2d = beluga::compose_interfaces<
     BaseParticleFilterInterface,
     StorageInterface<Sophus::SE2d, beluga::Weight>,
     EstimationInterface2d,
     OdometryMotionModelInterface2d,
     LaserSensorModelInterface2d<Map>,
     CustomExtensionMixinInterface>;
 
+/// \cond
+
+/// Wrapper to convert a motion model class into a mixin.
+template <typename Mixin, typename Model>
+struct MotionMixin : public Mixin, public Model {
+  using update_type = typename Model::update_type;
+
+  template <typename... Args>
+  constexpr explicit MotionMixin(Model model, Args&&... args)
+      : Model(std::move(model)), Mixin(static_cast<decltype(args)>(args)...) {}
+
+  void update_motion(const update_type& pose) { Model::update_motion(pose); }
```

Comment:
@nahueespinosa why method shadowing? Same below.

Reply:
Because this doesn't compile :smiling_face_with_tear:

```cpp
struct Interface {
    virtual void do_the_thing() = 0;
};

struct Base {
    void do_the_thing() {}
};

struct Derived : public Interface, public Base {};
```

And something like making models inherit from their specific interface doesn't work unless we use virtual inheritance. Which in turns breaks mixins since you can't perform static downcasts.

Reply:
Hmm, I'm confused. Why would we need virtual inheritance? If motion mixins inherit from motion models, and motion models implement a motion model interface, that then becomes the motion mixin interface as well, not directly inherited but obtained through a type constrain on the motion model class, there's no diamond problem. Or that's how I picture it.

Reply:
> If motion mixins inherit from motion models, and motion models implement a motion model interface, that then becomes the motion mixin interface as well

That doesn't work like that, the compiler doesn't piece together that the motion mixin inherited members are actually implementing pure interface members inherited by the whole mixin.

```cpp
struct Interface { virtual void do_the_thing() = 0; };

struct MotionModel { void do_the_thing() {} };

struct OtherMixins : public Interface {};

struct MotionMixin : public OtherMixins, public MotionModel {};

struct MostDerived : public MotionMixin {};
```

```bash
<source>:17:33: error: invalid cast to abstract class type 'MostDerived'
   17 |     auto instance = MostDerived{};
      |                                 ^
<source>:14:8: note:   because the following virtual functions are pure within 'MostDerived':
   14 | struct MostDerived : public MotionMixin {};
      |        ^~~~~~~~~~~
```

While doing this, I thought of this potential solution:
```cpp
struct Interface { virtual void do_the_thing() = 0; };

struct MotionModel : virtual public Interface { void do_the_thing() {} };

struct OtherMixins : virtual public Interface {};

struct MotionMixin : public OtherMixins, public MotionModel {};

struct MostDerived : public MotionMixin {};
```
But that breaks compile time static casts for self members.

Reply:
> That doesn't work like that, the compiler doesn't piece together that the motion mixin inherited members are actually implementing pure interface members inherited by the whole mixin.

Right. What I'm proposing is closer to your second example, on the basis that there will not be any mixins inheriting `Interface` directly, only through `MotionModel`. Thus virtual inheritance is not necessary.

Reply:
And then we cast the particle filter mixin pointer in the user code to the appropriate interface depending of if we want to update the sensor measurement or we want to _resample_ for instance?

Reply:
Hmm, but that will make it impossible to define a particle filter interface. Nevermind. You are right, there is no way around shadowing.

Reply:
Sadly right... The worst kind of right.

Reply:
I think it's not worth it. The future (ranges) looks greener and this is just a transition patch.

Reply:
Not even suggesting it (posts are getting interleaved :sweat_smile:). That'd be awful. This patch is OK as-is.).